### DISCUSSION TITLE
websocket_endpoint(websocket: WebSocket) why must I need the annotation?

### DISCUSSION THREAD

**Discussion by CSLukkun** [fastapi/fastapi#12461] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, WebSocket
from fastapi.responses import HTMLResponse

app = FastAPI()

html = """
<!DOCTYPE html>
<html>
    <head>
        <title>Chat</title>
    </head>
    <body>
        <h1>WebSocket Chat</h1>
        <form action="" onsubmit="sendMessage(event)">
            <input type="text" id="messageText" autocomplete="off"/>
            <button>Send</button>
        </form>
        <ul id='messages'>
        </ul>
        <script>
            var ws = new WebSocket("ws://localhost:8000/ws");
            ws.onmessage = function(event) {
                var messages = document.getElementById('messages')
                var message = document.createElement('li')
                var content = document.createTextNode(event.data)
                message.appendChild(content)
                messages.appendChild(message)
            };
            function sendMessage(event) {
                var input = document.getElementById("messageText")
                ws.send(input.value)
                input.value = ''
                event.preventDefault()
            }
        </script>
    </body>
</html>
"""


@app.get("/")
async def get():
    return HTMLResponse(html)


@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

### Description

I met a question and felt confused. I am not clear if it is bug.

The example code can run normally. while I delete the annotation in the function `websocket_endpoint`.
I find it cannot run and return 403 status code.
```
@app.websocket("/ws")
async def websocket_endpoint(websocket):
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

### Operating System

Linux

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

Python 3.10.14

**Comment by YuriiMotov** [comment, marked as answer]
FastAPI uses type annotations to identify internal objects like `Request`, `Response`, `BackgroundTasks`, `WebSocket` and others and pass them into your endpoints.
So that's why they are mandatory.

It's not mentioned in the documentation (at least I didn't find), but it's already discussed here.

Here is the part of source code related to this: https://github.com/fastapi/fastapi/blob/50639525bf0518a285544cae05eb4ad0b1fadb2e/fastapi/dependencies/utils.py#L309-L330

**Reply by CSLukkun** [reply]
Thanks for your answer.

Could you tell me how do you know that? by docs or source code. 

Thanks again
